### ISSUE TITLE
ECDSA Signature Error for NIP-26 event

### ISSUE THREAD

**Issue by dtonon** [mikedilger/gossip#375]
@catenocrypt I'm testing a NIP-26 implementation and Gossip gives me an error parsing the event https://www.nostr.guru/e/21553125277fa85be67a9339e5ede21ffe91567c258c124913ef79cbcdc770cb:

VERIFY ERROR: ECDSA Signature Error: signature error <event json>

I cannot understand why, the signature seems valid, idem for the delegation part.
I really don't know if the problem is related to NIP-26 but other events created with the same lib (mine) works well.
Any idea, please?

**Comment by catenocrypt** [comment]
I've found the source of the issue, there is a flaw in the code. `DelegationCondition` is stored as a struct, and it's compiled in a string form on the fly, in a canonical format.  This may change the string, which results in invalidating the signature.

In the sample message the condition string was
`kind=1&created_at<1686078180&created_at>1680807780`
but when it was re-created at signature check, it became
`kind=1&created_at>1680807780&created_at<1686078180`
where the order of the clauses is different, so the string is different, message hash will be different, so signature will not be correct.

**Comment by catenocrypt** [comment]
Fix is that `DelegationCondition` should preserve the string form when parsed from string. This is a fix in `nostr-types` crate.

Workaround is to use the clauses in the order used in the code, that is `kind`, `created_at>`, `created_at<`.
https://github.com/mikedilger/nostr-types/blob/master/src/types/delegation.rs#L39

**Comment by catenocrypt** [comment]
Fix PR in `nostr-types` https://github.com/mikedilger/nostr-types/pull/12
(fix will be in `gossip` only after PR is accepted, new lib version is created, and it is included in gossip)

**Comment by dtonon** [comment]
@catenocrypt thanks!

> but when it was re-created at signature check, it became
kind=1&created_at>1680807780&created_at<1686078180

I suspected that this was the culprit because was the only "strange" thing (this "inverted" condition is created by the nos2x extension) and in the past I created other delegated notes using my lib that worked just fine in Gossip.

**Comment by mikedilger** [comment]
I've merged into nostr-types. I also updated gossip with that revision (which required also updating gossip-relay-picker with the same rev).  This was all done on the unstable branch.

**Comment by catenocrypt** [comment]
Thanks for the fast action @mikedilger
